### ISSUE TITLE
关于预训练代码的问题

### ISSUE THREAD

**Issue by KyrieXDL** [Goldgaruda/Tianchi-NLP-News-Text-Classification-Rank-5-solution#1]
你好，我看到你预训练代码中只训练了mask lm任务，但是生成样本你还是会有两句话进行拼接，这个会有影响吗？还有你的sentence.txt的文件是怎么样的，每一行是怎么进行分句的？

**Comment by Goldgaruda** [comment]
标准的bert是mask 语言模型和Next Sentence Prediction多任务训练方式，但是也有研究表明，性能甚至会有影响。Bert的第一个任务是采用MaskLM方式来训练语言模型，通俗地来说就是输入一句话的时候，随机地选一些要预测的词，然后用一个特殊的符号[MASK]来替代它们，之后模型根据所给的标签去学习这些地方该填的词。第二个任务在双向语言模型的基础上额外增加了一个句子级别的连续性预测任务，即预测输入BERT的单文本是否为连续的文本，引入这个任务可以更好地让模型学到连续文本片段之间的关系。虽然这个任务后来被证明不如不加，反而影响了BERT的性能表现。所以在这次比赛中的next sentence 并没有用。

主要改动了run_pretraining.py这个代码，你可以对照谷歌官方的和我的看看区别。 我注释了NSP
https://github.com/google-research/bert/blob/master/run_pretraining.py

**Comment by KyrieXDL** [comment]
> 标准的bert是mask 语言模型和Next Sentence Prediction多任务训练方式，但是也有研究表明，性能甚至会有影响。Bert的第一个任务是采用MaskLM方式来训练语言模型，通俗地来说就是输入一句话的时候，随机地选一些要预测的词，然后用一个特殊的符号[MASK]来替代它们，之后模型根据所给的标签去学习这些地方该填的词。第二个任务在双向语言模型的基础上额外增加了一个句子级别的连续性预测任务，即预测输入BERT的单文本是否为连续的文本，引入这个任务可以更好地让模型学到连续文本片段之间的关系。虽然这个任务后来被证明不如不加，反而影响了BERT的性能表现。所以在这次比赛中的next sentence 并没有用。

你好，非常感谢你的回复！但是你生成样本还是保留了用于预测句子连续性的代码（就是对两句话中间加入[SEP]进行了拼接），这里会有影响吗？还有想了解一下你的sentence.txt每一行是怎么生成的呢？

**Comment by KyrieXDL** [comment]
> 标准的bert是mask 语言模型和Next Sentence Prediction多任务训练方式，但是也有研究表明，性能甚至会有影响。Bert的第一个任务是采用MaskLM方式来训练语言模型，通俗地来说就是输入一句话的时候，随机地选一些要预测的词，然后用一个特殊的符号[MASK]来替代它们，之后模型根据所给的标签去学习这些地方该填的词。第二个任务在双向语言模型的基础上额外增加了一个句子级别的连续性预测任务，即预测输入BERT的单文本是否为连续的文本，引入这个任务可以更好地让模型学到连续文本片段之间的关系。虽然这个任务后来被证明不如不加，反而影响了BERT的性能表现。所以在这次比赛中的next sentence 并没有用。
> 
> 主要改动了run_pretraining.py这个代码，你可以对照谷歌官方的和我的看看区别。 我注释了NSP
> https://github.com/google-research/bert/blob/master/run_pretraining.py

你好，我上面的问题就是关于create_pretraining_data.py这里面的代码

**Comment by Goldgaruda** [comment]
我把NSP这个loss注释掉了，所以并没有参与反向传播，主要loss还是来源于mask。所以实际跑的时候下一句的预测，loss很高。我并没有改log输出，所以还是会显示。因为平时bert分句有逗号，句号等标志，但是此题中是匿名化了，不是特别的清楚哪个数字代码代表分句的符号，（虽然可以通过统计概率猜出来），但是感觉意义不大理由如上，所以我就没有分句，而是按照预训练时候的固定长度分句的，最后nsp的性能并不好，我现在没找道当时的log，等有了，给你看。另外第一个是用了一个大bert，我机子跑不起来，所以就只能集成几个小bert来做，也和他们差不了太多。

关于加不加nsp预测，具体看实验，知乎也有讨论的
https://zhuanlan.zhihu.com/p/165893466
1.2.3 Next Sentence Prediction
用实验获得了和XLNet类似的结果，发现不加入Next Sentence Prediction(NSP)任务，直接用连续一长句训练效果更好
        # No NSP
        # (next_sentence_loss, next_sentence_example_loss,
        #  next_sentence_log_probs) = get_next_sentence_output(
        #     bert_config, model.get_pooled_output(), next_sentence_labels)
        #
        # total_loss = masked_lm_loss + next_sentence_loss

**Comment by KyrieXDL** [comment]
好的，非常感谢，最后再打扰一下问一下create_pretraining_data.py里的这个，就是如果只做mlm这个任务，那么生成训练样本这里是不是只用token_a或者token_b就好了，不用将他们再拼接起来?

`


        tokens.append("[CLS]") 
        segment_ids.append(0) 
        for token in tokens_a:
          tokens.append(token)
          segment_ids.append(0)

        tokens.append("[SEP]") 
        segment_ids.append(0)

        for token in tokens_b:
          tokens.append(token)
          segment_ids.append(1)
        tokens.append("[SEP]")
        segment_ids.append(1)

        (tokens, masked_lm_positions,
         masked_lm_labels) = create_masked_lm_predictions(
             tokens, masked_lm_prob, max_predictions_per_seq, vocab_words, rng)
        .......省略 
`

**Comment by Goldgaruda** [comment]
create_pretraining_data里面，加了一个函数，叫 create_instances_from_document_nsp的函数，和原本的create_instances_from_document区别是移除了NSP，因为你不知道句子的分隔符是哪个。
在 create_training_instances 中原本调用create_instances_from_document，现在调用的是create_instances_from_document_nsp

 Account for [CLS], [SEP], [SEP] --原来的
  max_num_tokens = max_seq_length - 3
-改动的
    # Account for [CLS], [SEP]
    max_segment_length = max_seq_length - 2

新的规则是根据你输入的最大长度，比如256来拆分一个文章，而不是之前的分割符号。
具体你对比下create_segments_from_document和create_instances_from_document_nsp。一个是新建的，一个是 重写了。
如下
def create_segments_from_document(document, max_segment_length):
    """Split single document to segments according to max_segment_length."""
    assert len(document) == 1
    document = document[0]
    document_len = len(document)

    index = list(range(0, document_len, max_segment_length)) #根据设定的最大长度拆分一段文章，间隔为最大的长度
    other_len = document_len % max_segment_length #末尾不能整除的部分
    if other_len > max_segment_length / 2: #大于一半的长度加入，否则舍去
        index.append(document_len)

    segments = []
    for i in range(len(index) - 1):
        segment = document[index[i]: index[i+1]]
        segments.append(segment) #把符合条件的拆分的小段加入训练的目标segments

    return segments

**Comment by KyrieXDL** [comment]
> create_pretraining_data里面，加了一个函数，叫 create_instances_from_document_nsp的函数，和原本的create_instances_from_document区别是移除了NSP，因为你不知道句子的分隔符是哪个。
> 在 create_training_instances 中原本调用create_instances_from_document，现在调用的是create_instances_from_document_nsp
> 
> Account for [CLS], [SEP], [SEP] --原来的
> max_num_tokens = max_seq_length - 3
> -改动的
> # Account for [CLS], [SEP]
> max_segment_length = max_seq_length - 2
> 
> 新的规则是根据你输入的最大长度，比如256来拆分一个文章，而不是之前的分割符号。
> 具体你对比下create_segments_from_document和create_instances_from_document_nsp。一个是新建的，一个是 重写了。
> 如下
> def create_segments_from_document(document, max_segment_length):
> """Split single document to segments according to max_segment_length."""
> assert len(document) == 1
> document = document[0]
> document_len = len(document)
> 
> ```
> index = list(range(0, document_len, max_segment_length)) #根据设定的最大长度拆分一段文章，间隔为最大的长度
> other_len = document_len % max_segment_length #末尾不能整除的部分
> if other_len > max_segment_length / 2: #大于一半的长度加入，否则舍去
>     index.append(document_len)
> 
> segments = []
> for i in range(len(index) - 1):
>     segment = document[index[i]: index[i+1]]
>     segments.append(segment) #把符合条件的拆分的小段加入训练的目标segments
> 
> return segments
> ```

非常感谢你的回复！开始看代码的不太仔细，感谢！！！

**Comment by KyrieXDL** [comment]
你好，又来打扰一下，请问你当时跑了大概多少个steps（我看到你跑了两次，一共60w steps？），最后的masked_lm_loss和masked_lm_accuracy大概多少呢，我感觉收敛的很慢。
